### PR TITLE
Add eligibility tags to home page

### LINKED ISSUE #4142: Show status on program list for in-progress application
**Description**
On the CiviForm programs page, show eligibility status (you may qualify, you may not qualify)

[Mocks](https://docs.google.com/presentation/d/1sNk1JSumJl8VbxKfukCkugIIm2zVDeNFWjJJuSWvflQ/edit#slide=id.g1bb302ee604_1_30)

**Additional context**
This is part of the eligibility work (#3744)

**Done when**
Eligibility status appears on program cards where eligibility questions have been answered

### PR DESCRIPTION
**Pull request by dkatzz**
### Description

Add tags whether an applicant may or may not be eligible for a program on the home page.

This is part of the eligibility work (#3744)

[Mocks](https://docs.google.com/presentation/d/1sNk1JSumJl8VbxKfukCkugIIm2zVDeNFWjJJuSWvflQ/edit#slide=id.g1bb302ee604_1_30)

## Release notes

The title of the pull request will be used as the default release notes description. If more detail is needed to communicate to partners the scope of the PR's changes, use this release notes section.

#### General

- [x] Added the correct label: < feature | bug | dependencies | infrastructure | ignore-for-release >
- [x] Created tests which fail without the change (if possible)
- [ ] Extended the README / documentation, if necessary

#### User visible changes

- [x] Wrote browser tests using the [validateAccessibility](https://sourcegraph.com/github.com/civiform/civiform/-/blob/browser-test/src/support/index.ts?L437:14&subtree=true) method
- [x] Manually tested at 200% size
- [ ] Manually evaluated tab order

#### New Features

- [ ] Add new FeatureFlag env vars to `server/conf/application.conf`
- [ ] Conditioned new functionality on a [FeatureFlag](https://docs.civiform.us/contributor-guide/developer-guide/feature-flags)
- [ ] Wrote browser tests with the feature flag off and on, etc.

### Issue(s) this completes

Fixes #4142

### CODE REVIEW

File: server/app/services/applicant/ApplicantService.java [hunk 1]
```diff
@@ -629,32 +646,110 @@ private ApplicationPrograms relevantProgramsForApplicant(
...
+                                (b) -> {
+                                  applicationProgramsFuture.complete(
+                                      appProgramsBuilder
+                                          .setUnapplied(sortByProgramId(unappliedPrograms.build()))
+                                          .build());
+                                });
+                  });
+    }
+    try {
+      return applicationProgramsFuture.get();
```
Comment (bion) [inline server/app/services/applicant/ApplicantService.java:747]: **issue:** here and a couple other places there are calls to `CompleteableFuture.get` which will block the current thread.

The code here is pretty complex, making it all the harder. This page is probably the biggest performance risk in the system.

I wonder if it's feasible to do more of the DB querying in the beginning and then collate the results synchronously without having to work with futures.

File: server/app/services/applicant/ApplicantService.java [hunk 2]
```diff
@@ -629,32 +646,110 @@ private ApplicationPrograms relevantProgramsForApplicant(
...
+                    .build());
+          });
+
+      applicationProgramsFuture.complete(
+          ApplicationPrograms.builder()
+              .setInProgress(sortByProgramId(inProgressPrograms.build()))
+              .setSubmitted(sortByProgramId(submittedPrograms.build()))
+              .setUnapplied(sortByProgramId(unappliedPrograms.build()))
+              .build());
+    } else {
```
Comment (bion) [inline server/app/services/applicant/ApplicantService.java:693]: **suggestion (blocking):** the if/else branches in this method have become pretty big making it hard to follow the flow of the code, if feasible please factor code into smaller methods

**Review by bion — CHANGES_REQUESTED** [review]
Please update unit test coverage for relevantProgramsForApplicant

File: browser-test/src/application_navigation.test.ts [hunk 3]
```diff
@@ -390,6 +391,24 @@ describe('Applicant navigation flow', () => {
...
+      await enableFeatureFlag(page, 'program_eligibility_conditions_enabled')
+      await applicantQuestions.applyProgram(fullProgramName)
+
+      // Fill out application and submit.
+      await applicantQuestions.answerNumberQuestion('5')
+      await applicantQuestions.clickNext()
+
+      // Verify the question is marked eligible
+      await applicantQuestions.gotoApplicantHomePage()
+      await applicantQuestions.seeEligibilityTag(fullProgramName, true)
```
[the remark this answers is not in the document]
Comment (dkatzz) [reply browser-test/src/application_navigation.test.ts]: Done.

File: browser-test/src/application_navigation.test.ts [hunk 4]
```diff
@@ -380,6 +380,7 @@ describe('Applicant navigation flow', () => {
 
       // Verify the question is marked ineligible.
       await applicantQuestions.gotoApplicantHomePage()
+      await applicantQuestions.seeEligibilityTag(fullProgramName, false)
```
[the remark this answers is not in the document]
Comment (dkatzz) [reply browser-test/src/application_navigation.test.ts]: Done.

File: browser-test/src/application_navigation.test.ts [hunk 5]
```diff
@@ -420,6 +439,8 @@ describe('Applicant navigation flow', () => {
 
       // Verify the question is marked ineligible.
       await applicantQuestions.gotoApplicantHomePage()
+      await validateScreenshot(page, 'ineligible-home-page-program-tag')
+      await applicantQuestions.seeEligibilityTag(fullProgramName, false)
```
[the remark this answers is not in the document]
Comment (dkatzz) [reply browser-test/src/application_navigation.test.ts]: Done.

File: browser-test/src/application_navigation.test.ts [hunk 6]
```diff
@@ -443,6 +464,7 @@ describe('Applicant navigation flow', () => {
 
       // Verify the question is marked ineligible.
       await applicantQuestions.gotoApplicantHomePage()
+      await applicantQuestions.seeEligibilityTag(fullProgramName, false)
```
[the remark this answers is not in the document]
Comment (dkatzz) [reply browser-test/src/application_navigation.test.ts]: Done.

File: server/app/services/applicant/ApplicantService.java (hunk 1 above)
Comment (dkatzz) [reply server/app/services/applicant/ApplicantService.java]: SO much better getting the data upfront - thanks for the suggestion. I didn't realize there were synchronous versions of the getReadOnlyApplicantProgramService function, but using those made it so that as long as I had the question data loaded in the program definitions, I could still initiate that class and use its methods.

File: server/app/services/applicant/ApplicantService.java (hunk 2 above)
Comment (dkatzz) [reply server/app/services/applicant/ApplicantService.java]: This is no longer applicable after refactoring and loading the data upfront.

File: server/app/services/applicant/ApplicantService.java [hunk 7]
```diff
@@ -167,6 +167,12 @@ public ReadOnlyApplicantProgramService getReadOnlyApplicantProgramService(
         application.getApplicantData(), programDefinition, baseUrl);
   }
 
+  /** Get a {@link ReadOnlyApplicantProgramService} from applicant data and a program definition. */
+  public ReadOnlyApplicantProgramService getReadOnlyApplicantProgramService(
```
Comment (bion) [inline server/app/services/applicant/ApplicantService.java:171]: **suggestion (non-blocking):** since this isn't used outside of this class consider making it `private`

File: server/app/services/applicant/ApplicantService.java (hunk 1 above)
Comment (bion) [reply server/app/services/applicant/ApplicantService.java]: Great work cleaning this up, I agree it's much improved!

File: server/app/services/applicant/ApplicantService.java (hunk 7 above)
Comment (dkatzz) [reply server/app/services/applicant/ApplicantService.java]: Good call, updated!

**Comment by dkatzz** [comment]
> Please update unit test coverage for relevantProgramsForApplicant

Updated!

File: server/app/services/program/ProgramService.java [hunk 8]
```diff
@@ -44,6 +44,18 @@ public interface ProgramService {
   /** Get the data object about the programs that are in the active or draft version. */
   ActiveAndDraftPrograms getActiveAndDraftPrograms();
 
+  /**
+   * Update all {@link QuestionDefinition}s in a list of {@link ProgramDefinition}s asynchronously,
```
Comment (bion) [inline server/app/services/program/ProgramService.java:48]: **nitpick (blocking):** `Update` here could be confused with a database update.
```suggestion
   * Sync all {@link QuestionDefinition}s in a list of {@link ProgramDefinition}s asynchronously,
```

File: server/app/services/program/ProgramService.java [hunk 9]
```diff
@@ -44,6 +44,18 @@ public interface ProgramService {
...
+   * by querying for questions then updating each {@link ProgramDefinition}s.
+   *
+   * @param programDefinitions the list of program definitions that should be updated
+   * @return a list of updated {@link ProgramDefinition}s with all of its associated questions if
+   *     they exist, or a QuestionNotFound is thrown when the future completes and a question is not
+   *     found.
```
Comment (bion) [inline server/app/services/program/ProgramService.java:54]: **suggestion (blocking):** use `@throws` instead of mentioning the error in the `@return` section

File: server/app/views/applicant/ProgramIndexView.java [hunk 10]
```diff
@@ -350,6 +367,21 @@ private PTag programCardApplicationStatus(
...
+    String color = isEligible ? "bg-blue-100" : "bg-gray-200";
+    String tagClass =
+        isEligible ? ReferenceClasses.ELIGIBLE_TAG : ReferenceClasses.NOT_ELIGIBLE_TAG;
+    String tagText =
+        isEligible
+            ? MessageKey.TAG_MAY_QUALIFY.getKeyName()
+            : MessageKey.TAG_MAY_NOT_QUALIFY.getKeyName();
+    return p().withClasses(tagClass, "border", "rounded-lg", "px-2", "py-1", "mb-4", color)
+        .with(
+            Icons.svg(icon).withClasses("inline-block").withStyle("width: 18px; height: 18px;"),
```
Comment (bion) [inline server/app/views/applicant/ProgramIndexView.java:381]: **question (blocking):** were the tailwind built-in classes not able to implement the design?

File: server/app/views/components/Icons.java [hunk 11]
```diff
@@ -29,6 +29,16 @@ public enum Icons {
       "M7 8h10M7 12h4m1 8l-4-4H5a2 2 0 01-2-2V6a2 2 0 012-2h14a2 2 0 012 2v8a2 2 0 01-2 2h-3l-4"
           + " 4z"),
   ARROW_LEFT(24, "m12 20-8-8 8-8 1.425 1.4-5.6 5.6H20v2H7.825l5.6 5.6Z"),
+  // Check Circle
```
Comment (bion) [inline server/app/views/components/Icons.java:32]: **suggestion (blocking):** Please add an external link to where we can see the new icon if feasible

File: server/app/views/components/Icons.java [hunk 12]
```diff
@@ -290,6 +300,16 @@ public enum Icons {
           + " 18c-4.41 0-8-3.59-8-8s3.59-8 8-8 8 3.59 8 8-3.59 8-8 8zm0-14c-2.21 0-4 1.79-4"
           + " 4h2c0-1.1.9-2 2-2s2 .9 2 2c0 2-3 1.75-3 5h2c0-2.25 3-2.5 3-5"
           + " 0-2.21-1.79-4-4-4z"),
+  INFO(
```
Comment (bion) [inline server/app/views/components/Icons.java:303]: **suggestion (blocking):** Please add an external link to where we can see the new icon if feasible

**Review by bion — CHANGES_REQUESTED** [review]

File: server/app/views/applicant/ProgramIndexView.java (hunk 10 above)
Comment (dkatzz) [reply server/app/views/applicant/ProgramIndexView.java:381]: Could you give more info? I implemented this similar to some Icons in ViewUtils (e.g. https://github.com/civiform/civiform/blob/main/server/app/views/ViewUtils.java#L107)

File: server/app/views/components/Icons.java (hunk 12 above)
Comment (dkatzz) [reply server/app/views/components/Icons.java:304]: Done!

File: server/app/views/components/Icons.java (hunk 11 above)
Comment (dkatzz) [reply server/app/views/components/Icons.java]: Done!

File: server/app/services/program/ProgramService.java (hunk 9 above)
Comment (dkatzz) [reply server/app/services/program/ProgramService.java]: Updated.

Comment (bion) [reply server/app/services/program/ProgramService.java]: Oh huh, yeah that's a bit of a surprise. Ah well, sorry for the noise

File: server/app/views/applicant/ProgramIndexView.java (hunk 10 above)
Comment (bion) [reply server/app/views/applicant/ProgramIndexView.java:381]: Ah that was in reference to using the inline styles `.withStyle("width: 18px; height: 18px;")` vs uses the tailwind width and height classes

Comment (dkatzz) [reply server/app/views/applicant/ProgramIndexView.java]: Ah makes sense! That height / width isn't supported - added a comment to make that clear.

**Review by bion — APPROVED** [review]
